### DISCUSSION TITLE
Surface Pro 9 Fedora 37 Touch Screen Not Working

### DISCUSSION THREAD

**Discussion by dj-jd-unofficial** [linux-surface/linux-surface#1042] [Q&A] [answered]
I installed fedora 37 on a surface pro 9, but cannot get the touch screen to work. (I want to get the surface pen to work as well, but i'm starting here.) The device is currently on _linux-surface_ kernel `6.1.6-1.surface.fc37.x86_64`. 

I first tried following [the main guide](https://github.com/linux-surface/linux-surface/wiki/Installation-and-Setup#fedora). `iptsd` does not work when trying to enable it. So I tried following @kkempter-getswift in
[ this thread](https://github.com/linux-surface/linux-surface/issues/987#issuecomment-1315828041). This is where i'm stuck. 

I installed the `dkms` module successfully. I unfortunately have multiple kernels to select from now, which might be part of my problem. I could not just run `sudo make dkms-install` in the directory where i downloaded [itch-linux](https://github.com/quo/ithc-linux). it gave me a hint to use `--kernelsourcedir` pointed to the correct directory of the extracted file. When running `sudo make dkms-install -kernelsourcedir /boot/vmlinuz-6.1.6-1.surface.fc37.x86_64`, i got a message at the end saying _make: Nothing to be done for '/boot/vmlinuz-6.1.6-1.surface.fc37.x86_64'._ When subsequently trying to run `sudo modprobe ithc`, i get _modprobe: FATAL: Module ithc not found in directory /lib/modules/6.1.6-1.surface.fc37.x86_64_ . I don't know what to do at this point.

Doing some research on this, i found an old article with a similar issue regarding a stylus pen driver. A reply in this thread said to [re-install DKMS linux and linux headers](https://ask.fedoraproject.org/t/dkms-kernel-headers-not-found/13770/8) because the headers can get corrupted on force shutdowns. Since my device is a 12th gen intel i had to force shut down my device plenty of times. Could this be part of the issue? I don't know how the linux kernel works so i can't say. 

Please help, i hope this thread will help others with their first fedora install on a surface pro 9 as well.

**Comment by StollD** [comment]
The guide you followed omits the most important step: `sudo dnf install kernel-surface-devel`

After that, `sudo make dkms-install` should work fine.

**Reply by dj-jd-unofficial** [reply]
Thank you for your reply @StollD ! 
I ran `sudo dnf install kernel-surface-devel`. It installed fine. Unfortunately, running `sudo make dkms-install` did not work. 
Here is what i got...

> [dj-jd@fedora ithc-linux-master]$ sudo make dkms-install
> [sudo] password for dj-jd: 
> test -e /usr/src/`sed -n '/^PACKAGE_NAME="\(.*\)"$/s//\1/p' dkms.conf`-`sed -n '/^PACKAGE_VERSION="\(.*\)"$/s//\1/p' dkms.conf` && make dkms-uninstall
> make[1]: Entering directory '/home/dj-jd/Downloads/ithc-linux-master'
> modprobe -r ithc
> modprobe: FATAL: Module ithc not found.
> make[1]: [Makefile:30: dkms-uninstall] Error 1 (ignored)
> dkms uninstall `sed -n '/^PACKAGE_NAME="\(.*\)"$/s//\1/p' dkms.conf`/`sed -n '/^PACKAGE_VERSION="\(.*\)"$/s//\1/p' dkms.conf`
> Module ithc 0.1 is not installed for kernel 6.1.6-1.surface.fc37.x86_64 (x86_64). Skipping...
> dkms remove `sed -n '/^PACKAGE_NAME="\(.*\)"$/s//\1/p' dkms.conf`/`sed -n '/^PACKAGE_VERSION="\(.*\)"$/s//\1/p' dkms.conf`
> Module ithc 0.1 is not installed for kernel 6.1.6-1.surface.fc37.x86_64 (x86_64). Skipping...
> Module ithc 0.1 is not built for kernel 6.1.6-1.surface.fc37.x86_64 (x86_64). Skipping...
> rm -rf /usr/src/`sed -n '/^PACKAGE_NAME="\(.*\)"$/s//\1/p' dkms.conf`-`sed -n '/^PACKAGE_VERSION="\(.*\)"$/s//\1/p' dkms.conf`
> sync
> make[1]: Leaving directory '/home/dj-jd/Downloads/ithc-linux-master'
> mkdir -p /usr/src/`sed -n '/^PACKAGE_NAME="\(.*\)"$/s//\1/p' dkms.conf`-`sed -n '/^PACKAGE_VERSION="\(.*\)"$/s//\1/p' dkms.conf`
> cp -r dkms.conf Makefile src /usr/src/`sed -n '/^PACKAGE_NAME="\(.*\)"$/s//\1/p' dkms.conf`-`sed -n '/^PACKAGE_VERSION="\(.*\)"$/s//\1/p' dkms.conf`
> dkms add `sed -n '/^PACKAGE_NAME="\(.*\)"$/s//\1/p' dkms.conf`/`sed -n '/^PACKAGE_VERSION="\(.*\)"$/s//\1/p' dkms.conf`
> Error! DKMS tree already contains: ithc-0.1
> You cannot add the same module/version combo more than once.
> make: *** [Makefile:24: dkms-install] Error 3

I am confused at the error at the beginning stating it cannot find `ithc`. It is in the source folder of the file i downloaded...

> [dj-jd@fedora ithc-linux-master]$ ls
> dkms.conf  Makefile  README.md  src
> [dj-jd@fedora ithc-linux-master]$ cd src
> [dj-jd@fedora src]$ ls
> ithc-debug.c  ithc-dma.h  ithc-main.c  ithc-regs.h  Kconfig
> ithc-dma.c    `ithc.h`      ithc-regs.c  Kbuild

Any thoughts on what to do next?

[ P.S your bio is wrong, I will hopefully use the software you wrote one day  :-)  ]

**Reply by StollD** [reply]
Try running `sudo make dkms-uninstall` and then `sudo rm -r /var/lib/dkms/ithc`. After that `sudo make dkms-install` again.

**Reply by dj-jd-unofficial** [reply]
Hi @StollD , thank you so much for your help, I'm very grateful for it. I think I'm closer, it looks like I was able to run:

1. `sudo make dkms-uninstall`
2. `sudo rm -r /var/lib/dkms/ithc`
3. and then rerun `sudo make dkms-install`

However, when i try to `sudo modprobe ithc` again, i get...
<pre>
modprobe: ERROR: could not insert &apos;ithc&apos;: Key was rejected by service
</pre>

Here is the output of my terminal from running all of the commands...
<pre>
[dj-jd@fedora ithc-linux-master]$ sudo make dkms-uninstall
[sudo] password for dj-jd: 

modprobe -r ithc
dkms uninstall `sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`/`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`
Module ithc-0.1 for kernel 6.1.6-1.surface.fc37.x86_64 (x86_64).
Before uninstall, this module version was ACTIVE on this kernel.

ithc.ko.xz:
 - Uninstallation
   - Deleting from: /lib/modules/6.1.6-1.surface.fc37.x86_64/extra/
 - Original module
   - No original module was found for this module on this kernel.
   - Use the dkms install command to reinstall any previous module version.
depmod....
dkms remove `sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`/`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`
Module ithc 0.1 is not installed for kernel 6.1.6-1.surface.fc37.x86_64 (x86_64). Skipping...
rm -rf /usr/src/`sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`-`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`
sync

[dj-jd@fedora ithc-linux-master]$ sudo rm -r /var/lib/dkms/ithc

[dj-jd@fedora ithc-linux-master]$ sudo make dkms-install

test -e /usr/src/`sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`-`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf` &amp;&amp; make dkms-uninstall
make: [Makefile:21: dkms-install] Error 1 (ignored)
mkdir -p /usr/src/`sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`-`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`
cp -r dkms.conf Makefile src /usr/src/`sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`-`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`
dkms add `sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`/`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`
Creating symlink /var/lib/dkms/ithc/0.1/source -&gt; /usr/src/ithc-0.1
dkms build `sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`/`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`
Sign command: /lib/modules/6.1.6-1.surface.fc37.x86_64/build/scripts/sign-file
Signing key: /var/lib/dkms/mok.key
Public certificate (MOK): /var/lib/dkms/mok.pub
Certificate or key are missing, generating self signed certificate for MOK...
Key file /var/lib/dkms/mok.key not found and can&apos;t be generated, modules won&apos;t be signed

Building module:
Cleaning build area...
make -j12 KERNELRELEASE=6.1.6-1.surface.fc37.x86_64 KDIR=/lib/modules/6.1.6-1.surface.fc37.x86_64/build...
Cleaning build area...
dkms install `sed -n &apos;/^PACKAGE_NAME=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`/`sed -n &apos;/^PACKAGE_VERSION=&quot;\(.*\)&quot;$/s//\1/p&apos; dkms.conf`

ithc.ko.xz:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/6.1.6-1.surface.fc37.x86_64/extra/
depmod....
sync

[dj-jd@fedora ithc-linux-master]$ sudo modprobe ithc

modprobe: ERROR: could not insert &apos;ithc&apos;: Key was rejected by service
</pre>

If the following information is useful or not, i don't know, but I was able to set the linux-surface kernel to be the default selection on startup via the wiki instructions. The surface pro 9 is set to boot from the middle setting (both Third-Party and "micro$oft only")

**Reply by StollD** [reply, marked as answer]
Modules need to be signed by a trusted key, otherwise the kernel will reject them. The easiest solution is to disable secureboot.

The slightly harder solution is to have DKMS sign the module automatically. For that you need to install openssl, and then reinstall the module. Once thats done you should have a key file at `/var/lib/dkms/mok.pub` that you have to enroll with mokutil. (https://ubuntu.com/blog/how-to-sign-things-for-secure-boot, section "Enrolling the key")

**Reply by dj-jd-unofficial** [reply]
Thank you all for your help
